chain/network/src/concurrency/scope/mod.rs

```diff
@@ -0,0 +1,459 @@
+        let (send, recv) = crossbeam_channel::bounded(1);
+        (Self { ctx, send }, recv)
+    }
+
+    pub fn send(&self, err: E) {
+        if let Ok(_) = self.send.try_send(err) {
+            self.ctx.cancel();
+        }
+    }
+}
+
+/// Internal representation of a scope.
+struct Inner<E: 'static> {
+    /// Signal sent once the scope is terminated (i.e. when Inner is dropped).
+    ///
+    /// Since all tasks keep a reference to the scope they belong to, all the tasks
+    /// of the scope are complete when terminated is sent.
+    terminated: signal::Once,
+    /// Context of this scope.
+    ///
+    /// It is a child context of the parent scope.
+    /// All tasks spawned in this scope are provided with this context.
+    ctx: ctx::Ctx,
+    /// The channel over which the first error reported by any completed task is sent.
+    ///
+    /// `output` channel has capacity 1, so that any subsequent attempts to send an error
+    /// will fail. The channel is created in `internal::run` and the value is received from
+    /// the channel only after the top-level scope has terminated.
+    /// Thanks to the fact that we have a single channel per top-level scope:
+    /// * if you await Service::terminate() from within some task, it will complete only
+    ///   after the error is actually registered, so the awaiting task won't be able to cause a
+    ///   race condition.
+    output: Output<E>,
+    /// Parent scope. We keep it to prevent termination of the parent, until this scope terminates.
+    _parent: Option<Arc<Self>>,
+}
+
+impl<E: 'static> Drop for Inner<E> {
+    fn drop(&mut self) {
+        self.terminated.send();
+    }
+}
+
+impl<E: 'static> Inner<E> {
+    /// Creates a new scope with the given context.
+    pub fn new(ctx: &ctx::Ctx) -> (Arc<Self>, crossbeam_channel::Receiver<E>) {
+        let ctx = ctx.sub(time::Deadline::Infinite);
+        let (output, recv) = Output::new(ctx.clone());
+        (Arc::new(Self { ctx, output, _parent: None, terminated: signal::Once::new() }), recv)
+    }
+
+    pub fn new_subscope(self: Arc<Self>) -> Arc<Inner<E>> {
+        Arc::new(Inner {
+            ctx: self.ctx.sub(time::Deadline::Infinite),
+            output: self.output.clone(),
+            _parent: Some(self),
+            terminated: signal::Once::new(),
+        })
+    }
+}
+
+/// Represents a task that can be joined by another task within Scope<'env>.
+/// We do not support awaiting for tasks outside of the scope, to simplify
+/// the concurrency model (you can still implement a workaround by using a channel,
+/// if you really want to, but that might mean that Scope is not what you want
+/// in the first place).
+/// In particular Service::spawn() doesn't return a JoinHandle,
+/// because it can be called outside of the scope that it belongs to.
+pub struct JoinHandle<'env, T>(
+    tokio::task::JoinHandle<Result<T, ErrTaskCanceled>>,
+    std::marker::PhantomData<fn(&'env ()) -> &'env ()>,
```

Comment:
It is invariant to make the reasoning easier. IMO ('env -> 'env) is a more readable way of expressing that sth is invariant (because it says explicitly that it is both covariant and contravariant) than "&'env mut" which requires understanding that a mutable reference actually equivalent to passing around a setter function, which is 'env -> ().